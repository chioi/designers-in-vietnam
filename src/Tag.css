.tag input {
  visibility: hidden;
  position: absolute;
}

.tag-content {
  padding: var(--space-small);
  z-index: 2;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.tag-content:hover {
  cursor: pointer;
}

.tag {
  display: flex;
  position: relative;
  font-size: var(--font-regular);
  margin: 0 var(--space-small);
  z-index: 3;
}

.tag canvas {
  z-index: 1;
  -webkit-transition: opacity 0.5s;
  -moz-transition: opacity 0.5s;
  -o-transition: opacity 0.5s;
  transition: opacity 0.5s;
  opacity: 0;
}

.tag--selected canvas {
  -webkit-transition: opacity 0.1s;
  -moz-transition: opacity 0.1s;
  -o-transition: opacity 0.1s;
  transition: opacity 0.1s;
  opacity: 1;
}
