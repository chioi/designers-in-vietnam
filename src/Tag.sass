@import "style/variables"
@import "style/mixins/space"
@import "style/mixins/typography"
@import "style/mixins/color"
@include colors($color)
@include padding-placeholders($space)
@include font-size-placeholders($font-sizes)
@include font-weight-placeholders($font-weights)

.tag
  @extend %font-tiny
  @extend %color-gray
  text-transform: uppercase
  display: flex
  position: relative
  z-index: 3

.tag.selected
  @extend %color-black
  @extend %font-bold

.tag.selected canvas
  transition: opacity 0.1s
  opacity: 1

.tag canvas
  transition: opacity 0.5s
  opacity: 0
  z-index: 1

.tag input
  visibility: hidden
  position: absolute

.tag-content
  @extend %padding-vertical-tiny
  @extend %padding-horizontal-regular
  user-select: none
  z-index: 2

.tag-content:hover
  cursor: pointer
