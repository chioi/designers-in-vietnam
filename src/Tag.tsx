import React, { memo, useCallback, useEffect, useRef } from "react";
import Brushstroke from "./brushstroke";
import "./Tag.css";

export interface ITag {
  id?: string;
  name: string;
}

export interface ITagProps {
  tag: ITag;
  onClick: (tag: ITag) => void;
  isSelected: boolean;
}

const getBoundingRect = (id: string) => {
  const boundingRect = document.getElementById(id);
  let height = 100;
  let width = 100;
  if (boundingRect) {
    boundingRect.getBoundingClientRect();
    height = boundingRect.offsetHeight;
    width = boundingRect.offsetWidth;
  }

  return { width, height };
};

const Tag = (props: ITagProps) => {
  const id = `${props.tag.name}-tag`;
  const ref = useRef(null);
  const bs: any = useRef(null);

  useEffect(() => {
    const { width, height } = getBoundingRect(id);
    if (ref.current) {
      const optionsBackground = {
        animation: "points",
        color: "#F5A623",
        duration: 0.1,
        height,
        inkAmount: 3,
        padding: 8,
        points: 4,
        queue: true,
        root: ref.current,
        size: 10,
        width
      };

      bs.current = new Brushstroke(optionsBackground);
    }
  }, []);

  useEffect(() => {
    if (!bs.current) {
      bs.current = new Brushstroke();
    }

    if (props.isSelected) {
      const { width, height } = getBoundingRect(id);
      bs.current.clear();
      bs.current.draw();
      bs.current.draw({
        animation: "points",
        color: "#F5A623",
        duration: 0.1,
        points: [12, height / 2, width - 12, height / 2]
      });
    }
  }, [props.isSelected]);

  const onClick = useCallback(() => {
    props.onClick(props.tag);
  }, []);

  return (
    <li ref={ref} className={`tag${props.isSelected ? " tag--selected" : ""}`}>
      <label
        id={id}
        htmlFor={`${props.tag.name}-input`}
        className="tag-content"
        onClick={onClick}
      >
        <input id={`${props.tag}-input`} type="checkbox" />
        {props.tag.name}
      </label>
    </li>
  );
};

export default memo(Tag);
