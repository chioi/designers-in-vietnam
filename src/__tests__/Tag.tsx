import React from "react";
import { act, create, ReactTestRenderer } from "react-test-renderer";
import Tag from "../Tag";

const tag = {
  id: "hs3l233",
  name: "ux"
};

const onClick = jest.fn();

describe("gets created", () => {
  let component: ReactTestRenderer;

  beforeAll(() => {
    component = create(<Tag tag={tag} onClick={onClick} isSelected={false} />);
  });

  it("matches the snapshot", () => {
    expect(component).toMatchSnapshot();
  });

  describe("gets clicked", () => {
    beforeAll(() => {
      component = create(<Tag tag={tag} onClick={onClick} isSelected={true} />);
      act(() => {
        component.root.findByType("label").props.onClick();
      });
    });

    it("calls the onClick function", () => {
      expect(onClick).toHaveBeenCalledTimes(1);
    });

    it("matches the snapshot", () => {
      expect(component).toMatchSnapshot();
    });
  });
});
