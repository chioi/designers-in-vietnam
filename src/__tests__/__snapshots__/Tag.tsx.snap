// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets created gets clicked matches the snapshot 1`] = `
<li
  className="tag tag--selected"
>
  <label
    className="tag-content"
    htmlFor="UX-input"
    id="UX-tag"
    onClick={[Function]}
  >
    <input
      id="[object Object]-input"
      type="checkbox"
    />
    UX
  </label>
</li>
`;

exports[`gets created matches the snapshot 1`] = `
<li
  className="tag"
>
  <label
    className="tag-content"
    htmlFor="UX-input"
    id="UX-tag"
    onClick={[Function]}
  >
    <input
      id="[object Object]-input"
      type="checkbox"
    />
    UX
  </label>
</li>
`;
